AppsMenu test components

labkey-chrisj opened the pull request:
#### Rationale
Adds component wrappers for the AppsMenu, which provides navigation between LKB/LKSM/LKS pages

#### Related Pull Requests
https://github.com/LabKey/biologics/pull/857  <- contains the tests that will use the components defined here

#### Changes
Because the menu-expand piece of this is different in LKSM/LKB than it is in LKS, there are two classes to wrap the base menu part- AppsMenu and LKAppsMenu.

Expanding the base menu shows the ProductsNavContainer, which looks like this:
![image](https://user-images.githubusercontent.com/16809856/114621524-a5104600-9c61-11eb-9e25-c9ac1355ddfa.png)

Clicking the "LabKey Server" link shows the LKSNavContainer, which looks like this: 
![image](https://user-images.githubusercontent.com/16809856/114621767-f4ef0d00-9c61-11eb-8f0d-d0216bed0d6a.png)

Clicking either SM or Biologics from the ProductsNavContainer shows the ProjectsNavContainer, which contains a list of projects of that kind and looks like this: 
![image](https://user-images.githubusercontent.com/16809856/114622057-58793a80-9c62-11eb-88eb-4adb68bca83b.png)

Clicking a project link in the ProjectsNavContainer shows the LeafNavContainer, which contains a list of product sections (shown here: a Biologics project) 
![image](https://user-images.githubusercontent.com/16809856/114622327-ac841f00-9c62-11eb-9757-258b8b10e4b5.png)

All of the NavContainer-named classes derive from BaseNavContainer.
Those with back-caret titles support navigating back to the previous context

labkey-chrisj committed:
wait for ready before interactions

labkey-chrisj committed:
add error/warning finder

labkey-chrisj committed:
Make the uploadPanel wait for interactive, also use fileUploadPanel

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_bio_media_sample_crud

labkey-chrisj committed:
add requested javadoc

labkey-chrisj committed:
fix for writeTsvContents(); actually write row contents

labkey-chrisj committed:
Update waitForLoaded on entityInsertPanel, surface fileUpload behavior and test it

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_entity_insert_pane_updates

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_entity_insert_pane_updates

labkey-chrisj committed:
modify waitForReady to support assay upload (with no target type select)

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	src/org/labkey/test/components/ui/entities/EntityInsertPanel.java
#	src/org/labkey/test/tests/component/EntityInsertPanelTest.java
#	src/org/labkey/test/util/TestDataGenerator.java

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into develop

labkey-chrisj committed:
Test coverage for AppsMenu component behavior

labkey-danield commented on src/org/labkey/test/components/ui/navigation/apps/LKAppsMenu.java:
```diff
@@ -0,0 +1,69 @@
+ * Wraps the expand/collapse toggle to show/hide the Apps-navigation menu in LabKey pages
+ * In SM or Biologics views, AppsMenu is the analog of this component
+ */
+public class LKAppsMenu extends BaseBootstrapMenu
```
It kind of feels like this could be merged with AppsMenu. If I understand correctly this is the "button" on the title bar, maybe if the locators in the getToggleLocator and in the finder used an 'or' condition you could combine these two files (`Locator.xpath("(//li[@id='headerProductDropdown']/a) | (//button[@id='product-navigation-button'])")`).

Maybe I'm missing something.

labkey-chrisj replied:
This is a good suggestion, I just worried that if I tried to get too fancy with those two parts you guys would ask me to separate them.
I'm willing to try to unify these two into one that can straddle both worlds, tho

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_AppsMenu_Test

labkey-chrisj committed:
C/R feedback:
-add re-routing tests within products
-consolidate LKAppsMenu/AppsMenu, delete LKAppsMenu
-test tab add/detection
-have navigateTo() handle the case where the project panel isn't shown due to it being the only project of that type
